File: webapp/go/chair_handlers.go
package main

import (
	"database/sql"
	"errors"
	"fmt"
	"net/http"
	"time"

	"github.com/oklog/ulid/v2"
)

type postChairRegisterRequest struct {
	Name  string `json:"name"`
	Model string `json:"model"`
}

type postChairRegisterResponse struct {
	AccessToken string `json:"access_token"`
	ID          string `json:"id"`
}

func chairPostRegister(w http.ResponseWriter, r *http.Request) {
	owner := r.Context().Value("owner").(*Owner)

	req := &postChairRegisterRequest{}
	if err := bindJSON(r, req); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	chairID := ulid.Make().String()

	if req.Name == "" || req.Model == "" {
		writeError(w, http.StatusBadRequest, errors.New("some of required fields(name, model) are empty"))
		return
	}

	accessToken := secureRandomStr(32)
	_, err := db.Exec(
		"INSERT INTO chairs (id, owner_id, name, model, is_active, access_token) VALUES (?, ?, ?, ?, ?, ?)",
		chairID, owner.ID, req.Name, req.Model, false, accessToken,
	)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	writeJSON(w, http.StatusCreated, &postChairRegisterResponse{
		AccessToken: accessToken,
		ID:          chairID,
	})
}

func chairPostActivate(w http.ResponseWriter, r *http.Request) {
	chair := r.Context().Value("chair").(*Chair)

	_, err := db.Exec("UPDATE chairs SET is_active = ? WHERE id = ?", true, chair.ID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func chairPostDeactivate(w http.ResponseWriter, r *http.Request) {
	chair := r.Context().Value("chair").(*Chair)

	_, err := db.Exec("UPDATE chairs SET is_active = ? WHERE id = ?", false, chair.ID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func chairPostCoordinate(w http.ResponseWriter, r *http.Request) {
	req := &Coordinate{}
	if err := bindJSON(r, req); err != nil {
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	chair := r.Context().Value("chair").(*Chair)

	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	chairLocationID := ulid.Make().String()
	defer tx.Rollback()
	if _, err := tx.Exec(
		`INSERT INTO chair_locations (id, chair_id, latitude, longitude) VALUES (?, ?, ?, ?)`,
		chairLocationID, chair.ID, req.Latitude, req.Longitude,
	); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	location := &ChairLocation{}
	if err := tx.Get(location, `SELECT * FROM chair_locations WHERE id = ?`, chairLocationID); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	rideRequest := &RideRequest{}
	if err := tx.Get(rideRequest, `SELECT * FROM ride_requests WHERE chair_id = ? ORDER BY updated_at DESC LIMIT 1`, chair.ID); err != nil {
		if !errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusInternalServerError, err)
			return
		}
	} else {
		status, err := getLatestRideRequestStatus(tx, rideRequest.ID)
		if err != nil {
			writeError(w, http.StatusInternalServerError, err)
			return
		}
		if status != "COMPLETED" && status != "CANCELED" {
			if req.Latitude == rideRequest.PickupLatitude && req.Longitude == rideRequest.PickupLongitude && status == "DISPATCHING" {
				if _, err := tx.Exec("INSERT INTO ride_request_statuses (id, ride_request_id, status) VALUES (?, ?, ?)", ulid.Make().String(), rideRequest.ID, "DISPATCHED"); err != nil {
					writeError(w, http.StatusInternalServerError, err)
					return
				}
			}

			if req.Latitude == rideRequest.DestinationLatitude && req.Longitude == rideRequest.DestinationLongitude && status == "CARRYING" {
				if _, err := tx.Exec("INSERT INTO ride_request_statuses (id, ride_request_id, status) VALUES (?, ?, ?)", ulid.Make().String(), rideRequest.ID, "ARRIVED"); err != nil {
					writeError(w, http.StatusInternalServerError, err)
					return
				}
			}
		}
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	writeJSON(w, http.StatusOK, map[string]interface{}{
		"datetime": location.CreatedAt,
	})
}

func chairGetNotification(w http.ResponseWriter, r *http.Request) {
	chair := r.Context().Value("chair").(*Chair)
	found := true
	rideRequest := &RideRequest{}
	status := ""
	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	defer tx.Rollback()

	if _, err := tx.Exec("SELECT * FROM chairs WHERE id = ? FOR UPDATE", chair.ID); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if err := tx.Get(rideRequest, `SELECT * FROM ride_requests WHERE chair_id = ? ORDER BY updated_at DESC LIMIT 1`, chair.ID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			found = false
		} else {
			writeError(w, http.StatusInternalServerError, err)
			return
		}
	}

	if found {
		status, err = getLatestRideRequestStatus(tx, rideRequest.ID)
		if err != nil {
			writeError(w, http.StatusInternalServerError, err)
			return
		}
	}

	if !found || status == "COMPLETED" || status == "CANCELED" {
		matchRequest := &RideRequest{}
		// TODO: いい感じに椅子とユーザーをマッチングさせる
		// MEMO: 多分距離と椅子の移動速度が関係しそう
		if err := tx.Get(matchRequest, `SELECT * FROM ride_requests WHERE chair_id IS NULL ORDER BY created_at LIMIT 1 FOR UPDATE`); err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				w.WriteHeader(http.StatusNoContent)
				return
			}
			writeError(w, http.StatusInternalServerError, err)
			return
		}

		if _, err := tx.Exec("UPDATE ride_requests SET chair_id = ? WHERE id = ?", chair.ID, matchRequest.ID); err != nil {
			writeError(w, http.StatusInternalServerError, err)
			return
		}

		if !found {
			rideRequest = matchRequest
			status = "MATCHING"
		}
	}

	user := &User{}
	err = tx.Get(user, "SELECT * FROM users WHERE id = ?", rideRequest.UserID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	writeJSON(w, http.StatusOK, &getChairRequestResponse{
		RequestID: rideRequest.ID,
		User: simpleUser{
			ID:   user.ID,
			Name: fmt.Sprintf("%s %s", user.Firstname, user.Lastname),
		},
		PickupCoordinate: Coordinate{
			Latitude:  rideRequest.PickupLatitude,
			Longitude: rideRequest.PickupLongitude,
		},
		DestinationCoordinate: Coordinate{
			Latitude:  rideRequest.DestinationLatitude,
			Longitude: rideRequest.DestinationLongitude,
		},
		Status: status,
	})
}

func chairGetNotificationSSE(w http.ResponseWriter, r *http.Request) {
	chair := r.Context().Value("chair").(*Chair)

	// Server Sent Events
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")

	var lastRideRequest *RideRequest
	var lastRideRequestStatus string
	for {
		select {
		case <-r.Context().Done():
			w.WriteHeader(http.StatusOK)
			return

		default:
			err := func() error {
				found := true
				rideRequest := &RideRequest{}
				tx, err := db.Beginx()
				if err != nil {
					return err
				}
				defer tx.Rollback()

				if _, err := tx.Exec("SELECT * FROM chairs WHERE id = ? FOR UPDATE", chair.ID); err != nil {
					return err
				}

				if err := tx.Get(rideRequest, `SELECT * FROM ride_requests WHERE chair_id = ? ORDER BY updated_at DESC LIMIT 1`, chair.ID); err != nil {
					if errors.Is(err, sql.ErrNoRows) {
						found = false
					} else {
						return err
					}
				}

				status, err := getLatestRideRequestStatus(tx, rideRequest.ID)
				if err != nil {
					return err
				}

				if !found || status == "COMPLETED" || status == "CANCELED" {
					matchRequest := &RideRequest{}
					// TODO: いい感じに椅子とユーザーをマッチングさせる
					// MEMO: 多分距離と椅子の移動速度が関係しそう
					if err := tx.Get(matchRequest, `SELECT * FROM ride_requests WHERE chair_id IS NULL ORDER BY created_at LIMIT 1 FOR UPDATE`); err != nil {
						if errors.Is(err, sql.ErrNoRows) {
							return nil
						}
						return err
					}

					if _, err := tx.Exec("UPDATE ride_requests SET chair_id = ? WHERE id = ?", chair.ID, matchRequest.ID); err != nil {
						return err
					}

					if !found {
						rideRequest = matchRequest
					}
				}

				if lastRideRequest != nil && rideRequest.ID == lastRideRequest.ID && status == lastRideRequestStatus {
					return nil
				}

				user := &User{}
				err = tx.Get(user, "SELECT * FROM users WHERE id = ?", rideRequest.UserID)
				if err != nil {
					return err
				}

				if err := tx.Commit(); err != nil {
					return err
				}

				if err := writeSSE(w, "matched", &getChairRequestResponse{
					RequestID: rideRequest.ID,
					User: simpleUser{
						ID:   user.ID,
						Name: fmt.Sprintf("%s %s", user.Firstname, user.Lastname),
					},
					PickupCoordinate: Coordinate{
						Latitude:  rideRequest.PickupLatitude,
						Longitude: rideRequest.PickupLongitude,
					},
					DestinationCoordinate: Coordinate{
						Latitude:  rideRequest.DestinationLatitude,
						Longitude: rideRequest.DestinationLongitude,
					},
					Status: status,
				}); err != nil {
					return err
				}
				lastRideRequest = rideRequest
				lastRideRequestStatus = status

				return nil
			}()

			if err != nil {
				writeError(w, http.StatusInternalServerError, err)
				return
			}

			time.Sleep(100 * time.Millisecond)
		}
	}
}

type simpleUser struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type getChairRequestResponse struct {
	RequestID             string     `json:"request_id"`
	User                  simpleUser `json:"user"`
	PickupCoordinate      Coordinate `json:"pickup_coordinate"`
	DestinationCoordinate Coordinate `json:"destination_coordinate"`
	Status                string     `json:"status"`
}

func chairGetRequest(w http.ResponseWriter, r *http.Request) {
	requestID := r.PathValue("request_id")

	rideRequest := &RideRequest{}
	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	defer tx.Rollback()

	if err := tx.Get(rideRequest, "SELECT * FROM ride_requests WHERE id = ?", requestID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, errors.New("request not found"))
			return
		}
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	status, err := getLatestRideRequestStatus(tx, rideRequest.ID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	user := &User{}
	if err := tx.Get(user, "SELECT * FROM users WHERE id = ?", rideRequest.UserID); err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	writeJSON(w, http.StatusOK, &getChairRequestResponse{
		RequestID: rideRequest.ID,
		User: simpleUser{
			ID:   user.ID,
			Name: fmt.Sprintf("%s %s", user.Firstname, user.Lastname),
		},
		PickupCoordinate: Coordinate{
			Latitude:  rideRequest.PickupLatitude,
			Longitude: rideRequest.PickupLongitude,
		},
		DestinationCoordinate: Coordinate{
			Latitude:  rideRequest.DestinationLatitude,
			Longitude: rideRequest.DestinationLongitude,
		},
		Status: status,
	})
}

func chairPostRequestAccept(w http.ResponseWriter, r *http.Request) {
	requestID := r.PathValue("request_id")

	chair := r.Context().Value("chair").(*Chair)

	rideRequest := &RideRequest{}
	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	defer tx.Rollback()

	if err := tx.Get(rideRequest, "SELECT * FROM ride_requests WHERE id = ?", requestID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, errors.New("request not found"))
			return
		}
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if rideRequest.ChairID.String != chair.ID {
		writeError(w, http.StatusBadRequest, errors.New("not assigned to this request"))
		return
	}

	if _, err := tx.Exec("INSERT INTO ride_request_statuses (id, ride_request_id, status) VALUES (?, ?, ?)", ulid.Make().String(), requestID, "DISPATCHING"); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func chairPostRequestDeny(w http.ResponseWriter, r *http.Request) {
	requestID := r.PathValue("request_id")

	chair := r.Context().Value("chair").(*Chair)

	rideRequest := &RideRequest{}
	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	defer tx.Rollback()

	if err := tx.Get(rideRequest, "SELECT * FROM ride_requests WHERE id = ? FOR UPDATE ", requestID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, errors.New("request not found"))
			return
		}
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if rideRequest.ChairID.String != chair.ID {
		writeError(w, http.StatusBadRequest, errors.New("not assigned to this request"))
		return
	}

	if _, err := tx.Exec("INSERT INTO ride_request_statuses (id, ride_request_id, status) VALUES (?, ?, ?)", ulid.Make().String(), requestID, "MATCHING"); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func chairPostRequestDepart(w http.ResponseWriter, r *http.Request) {
	requestID := r.PathValue("request_id")

	chair := r.Context().Value("chair").(*Chair)

	rideRequest := &RideRequest{}
	tx, err := db.Beginx()
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}
	defer tx.Rollback()

	if err := tx.Get(rideRequest, "SELECT * FROM ride_requests WHERE id = ? FOR UPDATE", requestID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			writeError(w, http.StatusNotFound, errors.New("request not found"))
			return
		}
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if rideRequest.ChairID.String != chair.ID {
		writeError(w, http.StatusBadRequest, errors.New("not assigned to this request"))
		return
	}

	status, err := getLatestRideRequestStatus(tx, rideRequest.ID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if status != "DISPATCHED" {
		writeError(w, http.StatusBadRequest, errors.New("chair has not arrived yet"))
		return
	}

	if _, err := tx.Exec("INSERT INTO ride_request_statuses (id, ride_request_id, status) VALUES (?, ?, ?)", ulid.Make().String(), requestID, "CARRYING"); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	if err := tx.Commit(); err != nil {
		writeError(w, http.StatusInternalServerError, err)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

File: webapp/csharp/ChairHandlers.cs
using System.Data;
using System.Text.Json;
using System.Text.Json.Serialization;
using Dapper;

public static class ChairHandlers
{
  private class ChairPostChairsRequest
  {
    [JsonPropertyName("name")] public string Name { get; init; } = string.Empty;

    [JsonPropertyName("model")] public string Model { get; init; } = string.Empty;

    [JsonPropertyName("chair_register_token")]
    public string ChairRegisterToken { get; init; } = string.Empty;
  }

  private class ChairPostChairsResponse
  {
    [JsonPropertyName("id")] public string Id { get; set; } = string.Empty;

    [JsonPropertyName("owner_id")] public string OwnerId { get; set; } = string.Empty;
  }

  public static async Task ChairPostChairsAsync(HttpContext context, IDbConnection db)
  {
    try
    {
      var request = await JsonSerializer.DeserializeAsync<ChairPostChairsRequest>(context.Request.Body);
      if (request == null || string.IsNullOrEmpty(request.Name) || string.IsNullOrEmpty(request.Model) ||
          string.IsNullOrEmpty(request.ChairRegisterToken))
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest,
          "Some of the required fields (name, model, chair_register_token) are empty.");
        return;
      }

      var owner = await db.QueryFirstOrDefaultAsync<Owner>(
        "SELECT * FROM owners WHERE chair_register_token = @ChairRegisterToken",
        new { request.ChairRegisterToken });
      if (owner == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status401Unauthorized, "Invalid chair_register_token.");
        return;
      }

      var chairId = Ulid.NewUlid().ToString();
      var accessToken = Utils.SecureRandomStr(32);

      await db.ExecuteAsync(
        "INSERT INTO chairs (id, owner_id, name, model, is_active, access_token) VALUES (@ID, @OwnerID, @Name, @Model, @IsActive, @AccessToken)",
        new
        {
          ID = chairId,
          OwnerID = owner.ID,
          Name = request.Name,
          Model = request.Model,
          IsActive = false,
          AccessToken = accessToken
        });

      context.Response.Cookies.Append("chair_session", accessToken, new CookieOptions
      {
        Path = "/",
        HttpOnly = true
      });

      var response = new ChairPostChairsResponse
      {
        Id = chairId,
        OwnerId = owner.ID
      };
      await Request.WriteJsonAsync(context, response, StatusCodes.Status201Created);
    }
    catch (Exception ex)
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, ex);
    }
  }

  private class PostChairActivityRequest
  {
    [JsonPropertyName("is_active")] public bool IsActive { get; init; }
  }

  public static async Task ChairPostActivityAsync(HttpContext context, IDbConnection db)
  {
    try
    {
      // `chair` が Context に格納されていると仮定
      if (context.Items["chair"] is not Chair chair)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status401Unauthorized, "Chair not found in context.");
        return;
      }

      // リクエストの解析
      var request = await JsonSerializer.DeserializeAsync<PostChairActivityRequest>(context.Request.Body);
      if (request == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Invalid request body.");
        return;
      }

      // データベース更新
      const string query = "UPDATE chairs SET is_active = @IsActive WHERE id = @ChairId";
      var rowsAffected = await db.ExecuteAsync(query, new { request.IsActive, ChairId = chair.ID });

      if (rowsAffected == 0)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Chair not found.");
        return;
      }

      // ステータスコード 204 No Content を返す
      context.Response.StatusCode = StatusCodes.Status204NoContent;
    }
    catch (Exception ex)
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, ex);
    }
  }

  private class Coordinate
  {
    [JsonPropertyName("latitude")] public int Latitude { get; init; }

    [JsonPropertyName("longitude")] public int Longitude { get; init; }
  }

  private class ChairPostCoordinateResponse
  {
    [JsonPropertyName("recorded_at")] public long RecordedAt { get; set; }
  }

  public static async Task ChairPostCoordinateAsync(HttpContext context, IDbConnection db)
  {
    try
    {
      if (context.Items["chair"] is not Chair chair)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status401Unauthorized, "Chair not found in context.");
        return;
      }

      var request = await JsonSerializer.DeserializeAsync<Coordinate>(context.Request.Body);
      if (request == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Invalid request body.");
        return;
      }

      using var transaction = db.BeginTransaction();

      var chairLocationId = Ulid.NewUlid().ToString();
      const string insertLocationQuery = """
                                         INSERT INTO chair_locations (id, chair_id, latitude, longitude)
                                         VALUES (@Id, @ChairId, @Latitude, @Longitude)
                                         """;
      await db.ExecuteAsync(insertLocationQuery, new
      {
        Id = chairLocationId,
        ChairId = chair.ID,
        request.Latitude,
        request.Longitude
      }, transaction);

      const string selectLocationQuery = "SELECT * FROM chair_locations WHERE id = @Id";
      var location =
        await db.QueryFirstOrDefaultAsync<ChairLocation>(selectLocationQuery, new { Id = chairLocationId },
          transaction);

      if (location == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError,
          "Failed to retrieve chair location.");
        return;
      }

      const string selectRideQuery = "SELECT * FROM rides WHERE chair_id = @ChairId ORDER BY updated_at DESC LIMIT 1";
      var ride = await db.QueryFirstOrDefaultAsync<Ride>(selectRideQuery, new { ChairId = chair.ID }, transaction);

      if (ride != null)
      {
        // ライドステータスの取得
        var status = await GetLatestRideStatusAsync(transaction, ride.ID);

        if (!string.IsNullOrEmpty(status) && status != "COMPLETED" && status != "CANCELED")
        {
          if (request.Latitude == ride.PickupLatitude &&
              request.Longitude == ride.PickupLongitude &&
              status == "ENROUTE")
          {
            const string insertPickupStatusQuery = """
                                                   INSERT INTO ride_statuses (id, ride_id, status)
                                                   VALUES (@Id, @RideId, 'PICKUP')
                                                   """;
            await db.ExecuteAsync(insertPickupStatusQuery, new
            {
              Id = Ulid.NewUlid().ToString(),
              RideId = ride.ID
            }, transaction);
          }

          if (request.Latitude == ride.DestinationLatitude &&
              request.Longitude == ride.DestinationLongitude &&
              status == "CARRYING")
          {
            const string insertArrivedStatusQuery = """
                                                    INSERT INTO ride_statuses (id, ride_id, status)
                                                    VALUES (@Id, @RideId, 'ARRIVED')
                                                    """;
            await db.ExecuteAsync(insertArrivedStatusQuery, new
            {
              Id = Ulid.NewUlid().ToString(),
              RideId = ride.ID
            }, transaction);
          }
        }
      }

      transaction.Commit();

      await Request.WriteJsonAsync(context, new ChairPostCoordinateResponse
      {
        RecordedAt = location.CreatedAt.ToUnixTimeMilliseconds()
      });
    }
    catch (Exception ex)
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, ex);
    }
  }

  private static async Task<string> GetLatestRideStatusAsync(IDbTransaction transaction, string rideId)
  {
    const string query = """
                         SELECT status FROM ride_statuses
                         WHERE ride_id = @RideId
                         ORDER BY created_at DESC LIMIT 1
                         """;
    return await transaction.Connection!.QueryFirstAsync<string>(query, new { RideId = rideId }, transaction);
  }

  private class ChairGetNotificationResponse
  {
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    [JsonPropertyName("data")]
    public ChairGetNotificationResponseData? Data { get; set; }

    [JsonPropertyName("retry_after_ms")] public int RetryAfterMs { get; set; }
  }

  private class ChairGetNotificationResponseData
  {
    [JsonPropertyName("ride_id")] public string RideId { get; set; } = string.Empty;

    [JsonPropertyName("user")] public SimpleUser User { get; set; } = new();

    [JsonPropertyName("pickup_coordinate")]
    public Coordinate PickupCoordinate { get; set; } = new();

    [JsonPropertyName("destination_coordinate")]
    public Coordinate DestinationCoordinate { get; set; } = new();

    [JsonPropertyName("status")] public string Status { get; set; } = string.Empty;
  }

  private class SimpleUser
  {
    [JsonPropertyName("id")] public string Id { get; set; } = string.Empty;

    [JsonPropertyName("name")] public string Name { get; set; } = string.Empty;
  }

  public static async Task ChairGetNotificationAsync(HttpContext context, IDbConnection db)
  {
    if (context.Items["chair"] is not Chair chair)
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, "Chair context is missing.");
      return;
    }

    using var transaction = db.BeginTransaction();

    try
    {
      var ride = await db.QueryFirstOrDefaultAsync<Ride>(
        "SELECT * FROM rides WHERE chair_id = @ChairId ORDER BY updated_at DESC LIMIT 1",
        new { ChairId = chair.ID },
        transaction
      );

      if (ride == null)
      {
        await Request.WriteJsonAsync(context, new ChairGetNotificationResponse { RetryAfterMs = 30 });
        return;
      }

      var yetSentRideStatus = await db.QueryFirstOrDefaultAsync<RideStatus>(
        "SELECT * FROM ride_statuses WHERE ride_id = @RideId AND chair_sent_at IS NULL ORDER BY created_at ASC LIMIT 1",
        new { RideId = ride.ID },
        transaction
      );

      string status;
      if (yetSentRideStatus != null)
      {
        status = yetSentRideStatus.Status;
        await db.ExecuteAsync(
          "UPDATE ride_statuses SET chair_sent_at = CURRENT_TIMESTAMP(6) WHERE id = @Id",
          new { yetSentRideStatus.ID },
          transaction
        );
      }
      else
      {
        status = await GetLatestRideStatusAsync(db, transaction, ride.ID);
      }

      var user = await db.QueryFirstOrDefaultAsync<User>(
        "SELECT * FROM users WHERE id = @UserId FOR SHARE",
        new { UserId = ride.UserID },
        transaction
      );

      if (user == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, "User not found.");
        return;
      }

      transaction.Commit();

      var response = new ChairGetNotificationResponse
      {
        Data = new ChairGetNotificationResponseData
        {
          RideId = ride.ID,
          User = new SimpleUser
          {
            Id = user.ID,
            Name = $"{user.Firstname} {user.Lastname}"
          },
          PickupCoordinate = new Coordinate
          {
            Latitude = ride.PickupLatitude,
            Longitude = ride.PickupLongitude
          },
          DestinationCoordinate = new Coordinate
          {
            Latitude = ride.DestinationLatitude,
            Longitude = ride.DestinationLongitude
          },
          Status = status
        },
        RetryAfterMs = 30
      };

      await Request.WriteJsonAsync(context, response);
    }
    catch (Exception ex)
    {
      transaction.Rollback();
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, ex);
    }
  }

  private static async Task<string> GetLatestRideStatusAsync(IDbConnection db, IDbTransaction transaction,
    string rideId)
  {
    return await db.QueryFirstAsync<string>(
      "SELECT status FROM ride_statuses WHERE ride_id = @RideId ORDER BY created_at DESC LIMIT 1",
      new { RideId = rideId },
      transaction
    );
  }

  private class PostChairRidesRideIDStatusRequest
  {
    [JsonPropertyName("status")] public string Status { get; init; } = string.Empty;
  }

  public static async Task ChairPostRideStatusAsync(HttpContext context, IDbConnection db)
  {
    var rideId = context.Request.RouteValues["ride_id"]?.ToString();
    if (string.IsNullOrEmpty(rideId))
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "ride_id is missing from the route.");
      return;
    }

    if (context.Items["chair"] is not Chair chair)
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, "Chair context is missing.");
      return;
    }

    var req = await JsonSerializer.DeserializeAsync<PostChairRidesRideIDStatusRequest>(context.Request.Body);
    if (req == null || string.IsNullOrEmpty(req.Status))
    {
      await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Invalid request body.");
      return;
    }

    using var transaction = db.BeginTransaction();

    try
    {
      var ride = await db.QueryFirstOrDefaultAsync<Ride>(
        "SELECT * FROM rides WHERE id = @RideId FOR UPDATE",
        new { RideId = rideId },
        transaction
      );

      if (ride == null)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status404NotFound, "Ride not found.");
        return;
      }

      if (ride.ChairID != chair.ID)
      {
        await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Not assigned to this ride.");
        return;
      }

      switch (req.Status.ToUpper())
      {
        case "ENROUTE":
          await db.ExecuteAsync(
            "INSERT INTO ride_statuses (id, ride_id, status) VALUES (@Id, @RideId, @Status)",
            new { Id = Ulid.NewUlid().ToString(), RideId = ride.ID, Status = "ENROUTE" },
            transaction
          );
          break;

        case "CARRYING":
          var latestStatus = await GetLatestRideStatusAsync(db, transaction, ride.ID);
          if (latestStatus != "PICKUP")
          {
            await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Chair has not arrived yet.");
            return;
          }

          await db.ExecuteAsync(
            "INSERT INTO ride_statuses (id, ride_id, status) VALUES (@Id, @RideId, @Status)",
            new { Id = Ulid.NewUlid().ToString(), RideId = ride.ID, Status = "CARRYING" },
            transaction
          );
          break;

        default:
          await Request.WriteErrorAsync(context, StatusCodes.Status400BadRequest, "Invalid status.");
          return;
      }

      transaction.Commit();
      context.Response.StatusCode = StatusCodes.Status204NoContent;
    }
    catch (Exception ex)
    {
      transaction.Rollback();
      await Request.WriteErrorAsync(context, StatusCodes.Status500InternalServerError, ex);
    }
  }
}
